vector/src/main/java/im/vector/app/features/spaces/explore/SpaceDirectoryViewModel.kt

```diff
@@ -0,0 +1,169 @@
+        }
+        observeJoinedRooms()
+        observeMembershipChanges()
+    }
+
+    private fun observeJoinedRooms() {
+        val queryParams = roomSummaryQueryParams {
+            memberships = listOf(Membership.JOIN)
+        }
+        session
+                .rx()
+                .liveRoomSummaries(queryParams)
+                .subscribe { list ->
+                    val joinedRoomIds = list
+                            ?.map { it.roomId }
+                            ?.toSet()
+                            ?: emptySet()
+
+                    setState {
+                        copy(joinedRoomsIds = joinedRoomIds)
+                    }
+                }
+                .disposeOnClear()
+    }
+
+    private fun observeMembershipChanges() {
+        session.rx()
+                .liveRoomChangeMembershipState()
+                .subscribe {
+                    setState { copy(changeMembershipStates = it) }
+                }
+                .disposeOnClear()
+    }
+
+    override fun handle(action: SpaceDirectoryViewAction) {
+        when (action) {
+            is SpaceDirectoryViewAction.ExploreSubSpace -> {
+                setState {
+                    copy(hierarchyStack = hierarchyStack + listOf(action.spaceChildInfo.childRoomId))
+                }
+            }
+            SpaceDirectoryViewAction.HandleBack -> {
+                withState {
+                    if (it.hierarchyStack.isEmpty()) {
+                        _viewEvents.post(SpaceDirectoryViewEvents.Dismiss)
+                    } else {
+                        setState {
+                            copy(
+                                    hierarchyStack = hierarchyStack.dropLast(1)
+                            )
+                        }
+                    }
+                }
+            }
+            is SpaceDirectoryViewAction.JoinOrOpen -> {
+                handleJoinOrOpen(action.spaceChildInfo)
+            }
+        }
+    }
+
+    private fun handleJoinOrOpen(spaceChildInfo: SpaceChildInfo) = withState { state ->
+        val isSpace = spaceChildInfo.roomType == RoomType.SPACE
+        if (state.joinedRoomsIds.contains(spaceChildInfo.childRoomId)) {
+            if (isSpace) {
+                handle(SpaceDirectoryViewAction.ExploreSubSpace(spaceChildInfo))
+            } else {
+                _viewEvents.post(SpaceDirectoryViewEvents.NavigateToRoom(spaceChildInfo.childRoomId))
+            }
+        } else {
+            // join
+            viewModelScope.launch {
+                try {
+                    if (isSpace) {
+                        session.spaceService().joinSpace(spaceChildInfo.childRoomId, null, spaceChildInfo.viaServers)
+                    } else {
+                        session.joinRoom(spaceChildInfo.childRoomId, null, spaceChildInfo.viaServers)
+                    }
+                } catch (failure: Throwable) {
+                    Timber.e(failure, "## Space: Failed to join room or subspace")
```

Comment:
notify the View?

Reply:
https://github.com/vector-im/element-android/issues/3207

Reply:
oki